### ISSUE TITLE
Linear 分类器中如何实现MaxCosine？

### ISSUE THREAD

**Issue by YaoMeng01** [Z-ZHHH/CVPR23-DML#4]
您好，首先感谢你完成如此优秀的项目并提供代码。我想问一下在Linear classifier下如何计算和使用MaxCosine做为未知类scoer？ 期待您的回复。

**Comment by Z-ZHHH** [comment]
您好，可以通过`F.normalize(out, dim=1) `操作计算余弦相似性输出，比如linear classifier丢弃bias项，对weight应用F.normalize，同时对模型的最后一层feature应用F.normalize，两个normalize后的tensor相乘计算余弦相似性，然后直接取其中的最大值为MaxCosine。

**Comment by YaoMeng01** [comment]
十分感谢您的回复，我想再请教一下，linear classifier 和 cosine classifier的区别是什么呢？ 我的理解是cosine classifier 也是分别对weight 和最后一层feature 进行normalize，然后对两个normalized后的tensor相乘输出余弦相似性。

**Comment by Z-ZHHH** [comment]
余弦分类器通过计算输入特征向量与每个类的均值向量之间的余弦相似度，将输入样本归类到相似度最高的类别。余弦分类器的作用主要体现在对特征向量的长度不敏感，在某些情况下对尺度变化表现出更好的鲁棒性。而linear分类器的范数是没有限制，可以看看论文[1]支撑材料里的图7。论文[2]也是比较早介绍这一方面探究的。

[1] Vaze S, Han K, Vedaldi A, et al. Open-set recognition: A good closed-set classifier is all you need?[J]. 2021.
[2] Ranjan R, Castillo C D, Chellappa R. L2-constrained softmax loss for discriminative face verification[J]. arXiv preprint arXiv:1703.09507, 2017.

**Comment by YaoMeng01** [comment]
> 您好，可以通过`F.normalize(out, dim=1) `操作计算余弦相似性输出，比如linear classifier丢弃bias项，对weight应用F.normalize，同时对模型的最后一层feature应用F.normalize，两个normalize后的tensor相乘计算余弦相似性，然后直接取其中的最大值为MaxCosine。 例如，`logit = net.fc(feature)` 改为
> 
> ```
> norm_w = F.normalize(net.fc.weight, dim=0)
> norm_feat = F.normalize(feature, dim=1)
> cosine = norm_feat @ norm_w.T
> ```
好的，我理解了，十分感谢，顺便再问一下，Linear classifier下对wight和feature的normalize和矩阵相乘计算是在测试阶段计算OODscore的时候进行的吧，在训练过程中不需要进行normalize操作对吧。

**Comment by Z-ZHHH** [comment]
嗯是的，在测试阶段进行，训练阶段还是正常的linear classifier
